Extensions/P2P/B_p2ptools.js

```diff
@@ -0,0 +1,308 @@
+/**
+ * Tools for p2p multiplayer.
+ * @namespace
+ */
+gdjs.evtTools.p2p = {
+  /**
+   * The peer to peer configuration.
+   */
+  peerConfig: {debug:1}, // Enable logging of critical errors
+
+  /**
+   * The p2p client.
+   * @type {?Peer}
+   */
+  peer: null,
+
+  /**
+   * All connected p2p clients, keyed by their id.
+   */
+  connections: {},
+
+  /**
+   * Contains a list of events triggered by other p2p clients.
+   */
+  triggeredEvents: {},
+
+  /**
+   * Contains the latest data sent with each event.
+   */
+  lastEventData: {},
+
+  /**
+   * Tells how to handle an event (with or without data loss)
+   */
+  eventHandling: {},
+
+  /**
+   * True if PeerJS is initialized and ready.
+   */
+  ready: false,
+
+  /**
+   * True if an error occured.
+   */
+  error: false,
+
+  /**
+   * Last error's message.
+   */
+  lastError: "",
+
+  /**
+   * True if a peer diconnected.
+   */
+  peerDisconnected: false,
+
+  /**
+   * The last peer that has disconnected.
+   */
+  disconnectedPeer: "",
+}
+
+gdjs.evtTools.p2p.loadPeerJS = function() {
+  if(gdjs.evtTools.p2p.peer != null) return;
+  gdjs.evtTools.p2p.peer = new Peer(gdjs.evtTools.p2p.peerConfig);
+  gdjs.evtTools.p2p.peer.on("open", function() {
+    gdjs.evtTools.p2p.ready = true;
+  });
+  gdjs.evtTools.p2p.peer.on("error", function(errorMessage) {
+    gdjs.evtTools.p2p.error = true;
+    gdjs.evtTools.p2p.lastError = errorMessage;
+  });
+  gdjs.evtTools.p2p.peer.on("connection", gdjs.evtTools.p2p._onConnection);
+  gdjs.evtTools.p2p.peer.on("close", function() {
+    gdjs.evtTools.p2p.peer = null;
+    gdjs.evtTools.p2p.loadPeerJS();
+  });
+  gdjs.evtTools.p2p.peer.on("disconnected", gdjs.evtTools.p2p.peer.reconnect);
+}
+
+gdjs.evtTools.p2p._onConnection = function(connection) {
+  gdjs.evtTools.p2p.connections[connection.peer] = connection;
+  connection.on("data", function(data) {
+    if(data.eventName === undefined) return;
+    var dataLoss = gdjs.evtTools.p2p.eventHandling[data.eventName];
+
+    if (typeof dataLoss === "undefined" || dataLoss === false) {
+      if(typeof gdjs.evtTools.p2p.lastEventData[data.eventName] !== "object")
+        gdjs.evtTools.p2p.lastEventData[data.eventName] = [];
+      gdjs.evtTools.p2p.lastEventData[data.eventName].push(data.data);
+    } else {
+      gdjs.evtTools.p2p.triggeredEvents[data.eventName] = true;
+      gdjs.evtTools.p2p.lastEventData[data.eventName] = data.data;
+    }
+  });
+  connection.on("error", function() {
+    // Close event is only for graceful disconnection, also handle error aka ungraceful disconnection
+    gdjs.evtTools.p2p._onDisconnect(connection.peer);
+  });
+  connection.on("close", function() {
+    gdjs.evtTools.p2p._onDisconnect(connection.peer);
+  });
+  // Regularly check for disconnection as the built in way is not reliable.
+  var disconnectChecker = function() {
+    if (connection.peerConnection.connectionState == "failed" || connection.peerConnection.connectionState == "disconnected") {
```

Comment:
Can you check everywhere that you used `===` instead of `==`? Basically, never ever used == in JS.

Reply:
I used it on purpose to check both for undefined or null

Reply:
For undefined it's fine then! But for strings/number/anything else, always go for === to avoid expensive typecastings and deoptimizations.